### ISSUE TITLE
The EnvironmentVariableSubstitutor in strict mode shouldn't throw if there's a default value

### ISSUE THREAD

**Issue by esneider** [dropwizard/dropwizard#2787]
Apache's `StringSubstitutor` allows to specify a default value for env var replacement, a la bash, in the form of `${MY_ENV_VAR:-default value}`.

Meanwhile, dropwizard's `EnvironmentVariableSubstitutor` has a strict mode where if an env var isn't defined, it will throw an `UndefinedEnvironmentVariableException` to avoid failing silently.

I would expect the `EnvironmentVariableSubstitutor` in strict mode to *not* fail if there's a missing env var, but there's a default value specified for it.

Right now the default value notation is useless if strict mode is enabled.

**Comment by nickbabcock** [comment]
I agree with your logic and should be fixed.

**Comment by nickbabcock** [comment]
Short rant.

There appears to be no extension point in the underlying commons-text [`StringSubstitutor`](https://github.com/apache/commons-text/blob/8b07e17402a7ca148aa6442c1c1e25a2f700c713/src/main/java/org/apache/commons/text/StringSubstitutor.java) where one can check the result after the default value is substituted.

I then tried inlining [the entire logic](https://github.com/apache/commons-text/blob/8b07e17402a7ca148aa6442c1c1e25a2f700c713/src/main/java/org/apache/commons/text/StringSubstitutor.java#L1320-L1457) and surgically inserting our desired behavior, but noticed a couple of issues that prevents a copy+paste implementation

1. `StringSubstitutor` uses the "package-protected for test code use only" `TextStringBuilder#buffer` field. So we'd need to use something like `TextStringBuidler#toCharArray()`
2. After using `TextStringBuidler#toCharArray()`, I noticed there is a bug in the commons-text implementation hiding `StringIndexOutOfBoundsException` as the `TextStringBuilder` internal buffer doesn't resize when a character is deleted, so when internal logic hits this snippet

```java
// on the first call initialize priorVariables
if (priorVariables == null) {
    priorVariables = new ArrayList<>();
    priorVariables.add(new String(chars, offset, length));
}
```

The issue is that `length` can be longer than `chars`, as `length` is never changed even when a character is deleted -- thus `new String` will throw an exception.

While I do have a solution that passed all of our tests, I'm unsure it is flawless so I'll need to work to file a bug report against commons-text and see if it passes their test suite. Ideally they'd allow an extension point, but I'm unsure if this is likely.

I was hoping this issue would be a quick fix, but looks like it won't be the case 😅
